api/resources/annotations.py

```diff
@@ -0,0 +1,42 @@
+from django.contrib.contenttypes.models import ContentType
+from django.db.models import Exists, IntegerField, OuterRef, Subquery
+
+from osf.models import OutcomeArtifact, Registration
+from osf.utils.outcomes import ArtifactTypes
+
+
+def _make_badge_subquery(artifact_type):
+    '''Make the subquery for a specific Open Practice Badge.
+
+    Requires `primary_outcome_id` annotation be present. As such, this should only be called
+    from make_open_practice_badge_annotations.
+    '''
+    return Exists(
+        OutcomeArtifact.objects.filter(
+            outcome_id=OuterRef('primary_outcome_id'),
+            finalized=True,
+            deleted__isnull=True,
+            artifact_type=artifact_type,
+        ),
+    )
+
+
+def make_open_practice_badge_annotations():
+    '''Builds all of the annotation Subqueries for OpenPractice badges. '''
+    #TODO: Accept CT as a param
+    primary_resource_ct = ContentType.objects.get_for_model(Registration)
+    primary_outcome_subquery = Subquery(
```

Comment:
Is this tested in any way, and if not, should it be?

Reply:
This annotation is used by the open badge annotations, so it is tested implicitly. The RegistrationList tests, in particular, demonstrate that this is correctly identifying the Outcome to filter on for a specific Registration.

I'll beef up the RegistrationDetail tests, though, to demonstrate that a Registration doesn't accidentally inherit badges by being a resource on a Registration that has earned them!